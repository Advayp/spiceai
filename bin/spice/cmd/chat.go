package cmd

import (
	"bufio"
	"bytes"
	"encoding/json"
	"fmt"
	"log/slog"
	"net/http"
	"os"
	"strings"
	"time"

	"github.com/manifoldco/promptui"
	"github.com/peterh/liner"
	"github.com/spf13/cobra"
	"github.com/spiceai/spiceai/bin/spice/pkg/api"
	"github.com/spiceai/spiceai/bin/spice/pkg/context"
	"github.com/spiceai/spiceai/bin/spice/pkg/util"
)

const (
	cloudKeyFlag        = "cloud"
	modelKeyFlag        = "model"
	httpEndpointKeyFlag = "http-endpoint"
	userAgentKeyFlag    = "user-agent"
)

type Message struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

type ChatRequestBody struct {
	Messages      []Message      `json:"messages"`
	Model         string         `json:"model"`
	Stream        bool           `json:"stream"`
	StreamOptions *StreamOptions `json:"stream_options"`
}

type StreamOptions struct {
	IncludeUsage bool `json:"include_usage"`
}

type Delta struct {
	Content      string      `json:"content"`
	FunctionCall interface{} `json:"function_call"`
	ToolCalls    interface{} `json:"tool_calls"`
	Role         interface{} `json:"role"`
}

type Choice struct {
	Index        int         `json:"index"`
	Delta        Delta       `json:"delta"`
	FinishReason interface{} `json:"finish_reason"`
	Logprobs     interface{} `json:"logprobs"`
}

type ChatCompletion struct {
	ID                string   `json:"id"`
	Choices           []Choice `json:"choices"`
	Created           int64    `json:"created"`
	Model             string   `json:"model"`
	SystemFingerprint string   `json:"system_fingerprint"`
	Object            string   `json:"object"`
	Usage             *Usage   `json:"usage"`
}

type Usage struct {
	CompletionTokens int `json:"completion_tokens"`
	PromptTokens     int `json:"prompt_tokens"`
	TotalTokens      int `json:"total_tokens"`
}

type OpenAIError struct {
	Message string `json:"message"`
}

type OpenAIErrorResponse struct {
	Error OpenAIError `json:"error"`
}

var chatCmd = &cobra.Command{
	Use:   "chat",
	Short: "Chat with the Spice.ai LLM agent",
	Example: `
# Start a chat session with local spiced instance
spice chat --model <model>

# Start a chat session with spiced instance in spice.ai cloud
spice chat --model <model> --cloud
`,
	Run: func(cmd *cobra.Command, args []string) {
		cloud, _ := cmd.Flags().GetBool(cloudKeyFlag)
		rtcontext := context.NewContext().WithCloud(cloud)
		err := rtcontext.Init()
		if err != nil {
			slog.Error("could not initialize runtime context", "error", err)
			os.Exit(1)
		}

		apiKey, _ := cmd.Flags().GetString("api-key")
		if apiKey != "" {
			rtcontext.SetApiKey(apiKey)
		}

		userAgent, _ := cmd.Flags().GetString(userAgentKeyFlag)
		if userAgent != "" {
			rtcontext.SetUserAgent(userAgent)
		} else {
			rtcontext.SetUserAgentClient("chat")
		}

		rtcontext.RequireModelsFlavor(cmd)

		model, err := cmd.Flags().GetString(modelKeyFlag)
		if err != nil {
			slog.Error("could not get model flag", "error", err)
			os.Exit(1)
		}
		if model == "" {
			models, err := api.GetData[api.Model](rtcontext, "/v1/models?status=true")
			if err != nil {
				slog.Error("could not list models", "error", err)
				os.Exit(1)
			}

			if len(models) == 0 {
				slog.Error("No models found")
				os.Exit(1)
			}

			availableModels := []string{}
			for _, model := range models {
				if model.Status == "Ready" {
					availableModels = append(availableModels, model.Name)
				}
			}

			if len(availableModels) == 0 {
				slog.Error("No models are ready")
				os.Exit(1)
			}

			selectedModel := availableModels[0]
			if len(availableModels) > 1 {

				prompt := promptui.Select{
					Label:        "Select model",
					Items:        availableModels,
					HideSelected: true,
				}

				_, selectedModel, err = prompt.Run()
				if err != nil {
					slog.Error("prompt failed", "error", err)
					return
				}
			}

			cmd.Printf("Using model: %s\n", selectedModel)
			model = selectedModel
		}

		httpEndpoint, err := cmd.Flags().GetString("http-endpoint")
		if err != nil {
			slog.Error("could not get http-endpoint flag", "error", err)
			os.Exit(1)
		}
		if httpEndpoint != "" {
			rtcontext.SetHttpEndpoint(httpEndpoint)
		}

		var messages []Message = []Message{}

		line := liner.NewLiner()
		line.SetCtrlCAborts(true)
		defer line.Close()
		for {
			message, err := line.Prompt("chat> ")
			if err == liner.ErrPromptAborted {
				break
			} else if err != nil {
				slog.Error("reading input line", "error", err)
				continue
			}

			line.AppendHistory(message)
			messages = append(messages, Message{Role: "user", Content: message})

			done := make(chan bool)
			go func() {
				util.ShowSpinner(done)
			}()

			body := &ChatRequestBody{
				Messages:      messages,
				Model:         model,
				Stream:        true,
				StreamOptions: &StreamOptions{IncludeUsage: true},
			}

			response, err := sendChatRequest(rtcontext, body)
			if err != nil {
				slog.Error("failed to send chat request to spiced", "error", err)
				continue
			}

			scanner := bufio.NewScanner(response.Body)
			var responseMessage = ""

			/// Usage for the entire stream, and related timing.
			var usage Usage
			startTime := time.Now()
			var endTime time.Time

			doneLoading := false

			for scanner.Scan() {
				chunk := scanner.Text()

				errorEvent, err := maybeErrorEvent(chunk, scanner)

				if err != nil {
					slog.Error("failed to decode error event", "error", err)
					continue
				}

				if errorEvent != nil {
					slog.Error("chat request failed", "error", errorEvent.Message)
					break
				}

				if !strings.HasPrefix(chunk, "data: ") {
					continue
				}
				chunk = strings.TrimPrefix(chunk, "data: ")

				var chatResponse ChatCompletion = ChatCompletion{}
				err = json.Unmarshal([]byte(chunk), &chatResponse)
				if err != nil {
					slog.Error("failed to unmarshal chat response", "error", err)
					continue
				}

				if !doneLoading {
					done <- true
					doneLoading = true
				}

				if chatResponse.Usage != nil {
					usage = *chatResponse.Usage
					endTime = time.Now()
				}

				if len(chatResponse.Choices) == 0 {
					continue
				}

				token := chatResponse.Choices[0].Delta.Content
				cmd.Printf("%s", token)
				responseMessage = responseMessage + token
			}

			if err := scanner.Err(); err != nil {
				slog.Error("error occurred while processing the input stream", "error", err)
			}

			if !doneLoading {
				done <- true
				doneLoading = true
			}

			if responseMessage != "" {
				messages = append(messages, Message{Role: "assistant", Content: responseMessage})
			}
			if usage != (Usage{}) {
				cmd.Printf("\n\n%s\n\n", generateUsageMessage(&usage, endTime.Sub(startTime).Abs()))
			} else {
				cmd.Print("\n\n")
			}
		}
	},
}

// `generateUsageMessage` generates a boxed summary of the usage statistics.
//
// ```shell
// +-----------------------------------------------------------+
// | Total tokens: 229 (53.60/s). Completion: 219. Prompt: 10. |
// +-----------------------------------------------------------+
// ```
func generateUsageMessage(u *Usage, dur time.Duration) string {
	if u == nil {
		return ""
	}

	tps := float32(dur.Milliseconds()) / 1000.0
	usage_line := fmt.Sprintf(
		"Total tokens: %d (%.2f/s). Completion: %d. Prompt: %d.",
		u.TotalTokens, tps, u.CompletionTokens, u.PromptTokens,
	)

	// Dynamically generate a line of the same length as the usage line
	line := strings.Repeat("-", len(usage_line)+2)
	return fmt.Sprintf("+%s+\n| %s |\n+%s+", line, usage_line, line)
}

func sendChatRequest(rtcontext *context.RuntimeContext, body *ChatRequestBody) (*http.Response, error) {
	jsonBody, err := json.Marshal(body)
	if err != nil {
		return nil, fmt.Errorf("error marshaling request body: %w", err)
	}

	url := fmt.Sprintf("%s/v1/chat/completions", rtcontext.HttpEndpoint())
	request, err := http.NewRequest("POST", url, bytes.NewReader(jsonBody))
	if err != nil {
		return nil, fmt.Errorf("error creating request: %w", err)
	}

	headers := rtcontext.GetHeaders()
	for key, value := range headers {
		request.Header.Set(key, value)
	}
	request.Header.Set("Content-Type", "application/json")

	response, err := rtcontext.Client().Do(request)
	if err != nil {
		return nil, fmt.Errorf("error sending request: %w", err)
	}

	return response, nil
}

func maybeErrorEvent(chunk string, scanner *bufio.Scanner) (*OpenAIError, error) {
	if strings.HasPrefix(chunk, "event: error") {
		scanner.Scan() // read line with error message
		errorMessage := scanner.Text()
		errorMessage = strings.TrimPrefix(errorMessage, "data: ")

		var errorResponse OpenAIErrorResponse = OpenAIErrorResponse{}
		err := json.Unmarshal([]byte(errorMessage), &errorResponse)
		if err != nil {
			return nil, fmt.Errorf("failed to unmarshal: %w", err)
		}

		return &errorResponse.Error, nil
	}

	return nil, nil
}

func init() {
	chatCmd.Flags().Bool(cloudKeyFlag, false, "Use cloud instance for chat (default: false)")
	chatCmd.Flags().String(modelKeyFlag, "", "Model to chat with")
	chatCmd.Flags().String(httpEndpointKeyFlag, "", "HTTP endpoint for chat (default: http://localhost:8090)")
	chatCmd.Flags().String(userAgentKeyFlag, "", "User agent to use in all requests")
	chatCmd.Flags().String("api-key", "", "The API key to use for authentication")

	RootCmd.AddCommand(chatCmd)
}
