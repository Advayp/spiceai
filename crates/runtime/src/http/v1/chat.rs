/*
Copyright 2024 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

use core::time;
use std::{convert::Infallible, sync::Arc, time::Duration};

use async_openai::types::{ChatCompletionResponseStream, CreateChatCompletionRequest};
use async_stream::stream;
use axum::{
    http::StatusCode,
    response::{
        sse::{Event, KeepAlive, Sse},
        IntoResponse, Response,
    },
    Extension, Json,
};
use futures::StreamExt;
use serde::Serialize;
use tokio::sync::RwLock;
use tracing::{Instrument, Span};

use crate::model::LLMModelStore;

pub(crate) async fn post(
    Extension(llms): Extension<Arc<RwLock<LLMModelStore>>>,
    Json(req): Json<CreateChatCompletionRequest>,
) -> Response {
    let span = tracing::span!(target: "task_history", tracing::Level::INFO, "ai_chat", input = %serde_json::to_string(&req).unwrap_or_default());
    span.in_scope(|| tracing::info!(target: "task_history", model = %req.model, "labels"));

    let span_clone = span.clone();
    async move {
        let model_id = req.model.clone();
        match llms.read().await.get(&model_id) {
            Some(model) => {
                if req.stream.unwrap_or_default() {
                    match model.chat_stream(req).await {
                        Ok(strm) => {
                            create_sse_response(strm, time::Duration::from_secs(30), span_clone)
                        }
                        Err(e) => {
                            tracing::error!(target: "task_history", parent: &span_clone, "{e}");
                            tracing::error!("Error from v1/chat: {e}");
                            StatusCode::INTERNAL_SERVER_ERROR.into_response()
                        }
                    }
                } else {
                    match model.chat_request(req).await {
                        Ok(response) => {
                            let preview = response
                                .choices
                                .first()
                                .map(|s| serde_json::to_string(s).unwrap_or_default())
                                .unwrap_or_default();

                            tracing::info!(target: "task_history", parent: &span_clone, captured_output = %preview);
                            Json(response).into_response()
                        }
                        Err(e) => {
                            tracing::error!(target: "task_history", parent: &span_clone, "{e}");
                            tracing::error!("Error from v1/chat: {e}");
                            StatusCode::INTERNAL_SERVER_ERROR.into_response()
                        }
                    }
                }
            }
            None => StatusCode::NOT_FOUND.into_response(),
        }
    }
    .instrument(span)
    .await
}

/// Create a SSE [`axum::response::Response`] from a [`ChatCompletionResponseStream`].
fn create_sse_response(
    mut strm: ChatCompletionResponseStream,
    keep_alive_interval: Duration,
    span: Span,
) -> Response {
    Sse::new(Box::pin(stream! {
        let mut chat_output = String::new();
        while let Some(msg) = strm.next().instrument(span.clone()).await {
            match msg {
                Ok(resp) => {
                    if let Some(choice) = resp.choices.first() {
                        if let Some(intermediate_chat_output) = &choice.delta.content {
                            chat_output.push_str(intermediate_chat_output);
                        }
                    }

                    yield Ok::<Event, Infallible>(Event::default().json_data(resp).unwrap_or_else(|e| {
                        tracing::error!("Failed to serialize chat completion message: {e}");
                        to_openai_error_event(e.to_string())
                    }));
                },
                Err(e) => {
                    tracing::error!("Error encountered in chat completion stream: {e}");
                    yield Ok(to_openai_error_event(e.to_string()));
                    break;
                }
            }
        };
        tracing::info!(target: "task_history", parent: &span, captured_output = %chat_output);
        drop(span);
    }))
    .keep_alive(KeepAlive::new().interval(keep_alive_interval))
    .into_response()
}

/// Create an [`Event`] that corresponds to an `OpenAI` error event.
///
/// `https://platform.openai.com/docs/api-reference/realtime-server-events/error`
fn to_openai_error_event(err: impl Into<String>) -> Event {
    Event::default()
        .event("error")
        .json_data(OpenaiErrorEvent::new(err))
        .unwrap_or_default()
}

#[derive(Serialize)]
pub struct ApiError {
    message: String,
}

#[derive(Serialize)]
pub struct OpenaiErrorEvent {
    r#type: String,
    error: ApiError,
}

impl OpenaiErrorEvent {
    pub fn new(err: impl Into<String>) -> Self {
        Self {
            r#type: "error".to_string(),
            error: ApiError {
                message: err.into(),
            },
        }
    }
}
