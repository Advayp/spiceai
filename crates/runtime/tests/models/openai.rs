/*
Copyright 2024 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

use app::AppBuilder;
use async_openai::types::{
    ChatCompletionRequestSystemMessageArgs, ChatCompletionRequestUserMessageArgs,
    CreateChatCompletionRequestArgs, EmbeddingInput,
};
use std::sync::Arc;

use llms::chat::Chat;
use opentelemetry_sdk::trace::TracerProvider;
use runtime::{auth::EndpointAuth, model::try_to_chat_model, Runtime};
use spicepod::component::{
    embeddings::{ColumnEmbeddingConfig, Embeddings},
    model::Model,
};

use crate::{
    init_tracing, init_tracing_with_task_history,
    models::{
        create_api_bindings_config, get_executed_tasks, get_params_with_secrets,
        get_taxi_trips_dataset, get_tpcds_dataset, json_is_single_row_with_value,
        normalize_chat_completion_response, normalize_embeddings_response,
        normalize_search_response, pretty_json_str, send_chat_completions_request,
        send_nsql_request, send_search_request,
    },
    utils::{runtime_ready_check, verify_env_secret_exists},
};

use super::send_embeddings_request;

#[tokio::test]
async fn openai_test_nsql() -> Result<(), anyhow::Error> {
    let _tracing = init_tracing(None);

    verify_env_secret_exists("SPICE_OPENAI_API_KEY")
        .await
        .map_err(anyhow::Error::msg)?;

    let app = AppBuilder::new("text-to-sql")
        .with_dataset(get_taxi_trips_dataset())
        .with_model(get_openai_model("gpt-4o-mini", "nql"))
        .with_model(get_openai_model("gpt-4o-mini", "nql-2"))
        .build();

    let api_config = create_api_bindings_config();
    let http_base_url = format!("http://{}", api_config.http_bind_address);

    let rt = Arc::new(Runtime::builder().with_app(app).build().await);

    let (_tracing, trace_provider) = init_tracing_with_task_history(None, &rt);

    let rt_ref_copy = Arc::clone(&rt);
    tokio::spawn(async move {
        Box::pin(rt_ref_copy.start_servers(api_config, None, EndpointAuth::no_auth())).await
    });

    tokio::select! {
        () = tokio::time::sleep(std::time::Duration::from_secs(10)) => {
            return Err(anyhow::anyhow!("Timed out waiting for components to load"));
        }
        () = rt.load_components() => {}
    }

    runtime_ready_check(&rt).await;

    // example responses for the test query: '[{"count(*)":10}]', '[{"record_count":10}]'

    tracing::info!("/v1/nsql: Verify default nsql request");
    let task_start_time = std::time::SystemTime::now();

    let response = send_nsql_request(
        http_base_url.as_str(),
        "how many records in taxi_trips dataset?",
        None,
        Some(false),
        None,
    )
    .await?;

    assert!(
        json_is_single_row_with_value(&response, 10),
        "Expected a single record containing the value 10"
    );

    // ensure all spans are exported into task_history
    let _ = trace_provider.force_flush();

    let tasks = get_executed_tasks(&rt, task_start_time.into()).await?;

    let table_schema_task = tasks
        .iter()
        .find(|t| t.0 == "tool_use::table_schema")
        .expect("Expected 'tool_use::table_schema' task to be executed");

    insta::assert_snapshot!(
        "nsql_table_schema_task",
        pretty_json_str(&table_schema_task.1)?
    );

    tracing::info!("/v1/nsql: Ensure model selection works");
    let response = send_nsql_request(
        http_base_url.as_str(),
        "how many records in taxi_trips dataset?",
        Some("nql-2"),
        Some(false),
        None,
    )
    .await?;

    assert!(
        json_is_single_row_with_value(&response, 10),
        "Expected a single record containing the value 10"
    );

    tracing::info!("/v1/nsql: Verify nsql request with 'sample_data_enabled:true'");

    let task_start_time = std::time::SystemTime::now();

    let response = send_nsql_request(
        http_base_url.as_str(),
        "how many records in taxi_trips dataset?",
        Some("nql"),
        Some(true),
        None,
    )
    .await?;

    assert!(
        json_is_single_row_with_value(&response, 10),
        "Expected a single record containing the value 10"
    );

    // ensure all spans are exported into task_history
    let _ = trace_provider.force_flush();

    let tasks = get_executed_tasks(&rt, task_start_time.into()).await?;

    let sample_data_task = tasks
        .iter()
        .find(|t| t.0 == "tool_use::sample_data")
        .expect("Expected 'tool_use::sample_data' task to be executed");

    insta::assert_snapshot!("nsql_sample_data_task", sample_data_task.1);

    let sql_query_task = tasks
        .iter()
        .find(|t| t.0 == "sql_query")
        .expect("Expected 'sql_query' task to be executed");

    insta::assert_snapshot!("nsql_sample_data_task_sql_query", sql_query_task.1);

    Ok(())
}

#[tokio::test]
async fn openai_test_search() -> Result<(), anyhow::Error> {
    let _tracing = init_tracing(None);

    verify_env_secret_exists("SPICE_OPENAI_API_KEY")
        .await
        .map_err(anyhow::Error::msg)?;

    let mut ds_tpcds_item = get_tpcds_dataset("item");
    ds_tpcds_item.embeddings = vec![ColumnEmbeddingConfig {
        column: "i_item_desc".to_string(),
        model: "openai_embeddings".to_string(),
        primary_keys: Some(vec!["i_item_sk".to_string()]),
        chunking: None,
    }];

    let app = AppBuilder::new("search_app")
        // taxi_trips dataset is used to test search when there is a dataset w/o embeddings
        .with_dataset(get_taxi_trips_dataset())
        .with_dataset(ds_tpcds_item)
        .with_embedding(get_openai_embeddings(
            Option::<String>::None,
            "openai_embeddings",
        ))
        .build();

    let api_config = create_api_bindings_config();
    let http_base_url = format!("http://{}", api_config.http_bind_address);
    let rt = Arc::new(Runtime::builder().with_app(app).build().await);

    let rt_ref_copy = Arc::clone(&rt);
    tokio::spawn(async move {
        Box::pin(rt_ref_copy.start_servers(api_config, None, EndpointAuth::no_auth())).await
    });

    tokio::select! {
        () = tokio::time::sleep(std::time::Duration::from_secs(10)) => {
            return Err(anyhow::anyhow!("Timed out waiting for components to load"));
        }
        () = rt.load_components() => {}
    }

    runtime_ready_check(&rt).await;

    tracing::info!("/v1/search: Ensure simple search request succeeds");
    let response = send_search_request(
        http_base_url.as_str(),
        "new patient",
        Some(2),
        Some(vec!["item".to_string()]),
        None,
        Some(vec!["i_color".to_string(), "i_item_id".to_string()]),
    )
    .await?;

    insta::assert_snapshot!(format!("search_1"), normalize_search_response(response));

    tracing::info!("/v1/search: Ensure search request across all datasets succeeds");
    let response = send_search_request(
        http_base_url.as_str(),
        "new patient",
        Some(2),
        None,
        None,
        None,
    )
    .await?;

    insta::assert_snapshot!(format!("search_2"), normalize_search_response(response));

    Ok(())
}

#[tokio::test]
async fn openai_test_embeddings() -> Result<(), anyhow::Error> {
    let _tracing = init_tracing(None);

    verify_env_secret_exists("SPICE_OPENAI_API_KEY")
        .await
        .map_err(anyhow::Error::msg)?;

    let app = AppBuilder::new("search_app")
        .with_embedding(get_openai_embeddings(
            Some("text-embedding-3-small"),
            "openai_embeddings",
        ))
        .build();

    let api_config = create_api_bindings_config();
    let http_base_url = format!("http://{}", api_config.http_bind_address);
    let rt = Arc::new(Runtime::builder().with_app(app).build().await);

    let rt_ref_copy = Arc::clone(&rt);
    tokio::spawn(async move {
        Box::pin(rt_ref_copy.start_servers(api_config, None, EndpointAuth::no_auth())).await
    });

    tokio::select! {
        () = tokio::time::sleep(std::time::Duration::from_secs(10)) => {
            return Err(anyhow::anyhow!("Timed out waiting for components to load"));
        }
        () = rt.load_components() => {}
    }

    runtime_ready_check(&rt).await;

    let embeddins_test = vec![
        (
            EmbeddingInput::String("The food was delicious and the waiter...".to_string()),
            Some("float"),
            None,
            None,
        ),
        (
            EmbeddingInput::StringArray(vec![
                "The food was delicious".to_string(),
                "and the waiter...".to_string(),
            ]),
            None,
            Some("test_user_id"),
            Some(256),
        ),
        (
            EmbeddingInput::StringArray(vec![
                "The food was delicious".to_string(),
                "and the waiter...".to_string(),
            ]),
            Some("base64"),
            None,
            Some(128),
        ),
    ];

    let mut test_id = 0;

    for (input, encoding_format, user, dimensions) in embeddins_test {
        test_id += 1;
        let response = send_embeddings_request(
            http_base_url.as_str(),
            "openai_embeddings",
            input,
            encoding_format,
            user,
            dimensions,
        )
        .await?;

        insta::assert_snapshot!(
            format!("embeddings_{}", test_id),
            // OpenAI's embeddings response is not deterministic (values vary for the same input, model version, and parameters) so
            // we normalize the response before snapshotting
            normalize_embeddings_response(response)
        );
    }

    Ok(())
}

#[tokio::test]
async fn openai_test_chat_completion() -> Result<(), anyhow::Error> {
    let _tracing = init_tracing(None);

    verify_env_secret_exists("SPICE_OPENAI_API_KEY")
        .await
        .map_err(anyhow::Error::msg)?;

    let mut model_with_tools = get_openai_model("gpt-4o-mini", "openai_model");
    model_with_tools
        .params
        .insert("tools".to_string(), "auto".into());

    let app = AppBuilder::new("text-to-sql")
        .with_dataset(get_taxi_trips_dataset())
        .with_model(model_with_tools)
        .build();

    let api_config = create_api_bindings_config();
    let http_base_url = format!("http://{}", api_config.http_bind_address);
    let rt = Arc::new(Runtime::builder().with_app(app).build().await);

    let rt_ref_copy = Arc::clone(&rt);
    tokio::spawn(async move {
        Box::pin(rt_ref_copy.start_servers(api_config, None, EndpointAuth::no_auth())).await
    });

    tokio::select! {
        () = tokio::time::sleep(std::time::Duration::from_secs(10)) => {
            return Err(anyhow::anyhow!("Timed out waiting for components to load"));
        }
        () = rt.load_components() => {}
    }

    runtime_ready_check(&rt).await;

    let response = send_chat_completions_request(
        http_base_url.as_str(),
        vec![
            ("system".to_string(), "You are an assistant that responds to queries by providing only the requested data values without extra explanation.".to_string()),
            ("user".to_string(), "Provide the total number of records in the taxi trips dataset. If known, return a single numeric value.".to_string()),
        ],
        "openai_model",
        false,
    ).await?;

    insta::assert_snapshot!(
        "chat_completion",
        normalize_chat_completion_response(response, false)
    );

    Ok(())
}

#[tokio::test]
async fn openai_test_chat_messages() -> Result<(), anyhow::Error> {
    let _tracing = init_tracing(None);

    verify_env_secret_exists("SPICE_OPENAI_API_KEY")
        .await
        .map_err(anyhow::Error::msg)?;

    let mut ds_tpcds_item = get_tpcds_dataset("item");
    ds_tpcds_item.embeddings = vec![ColumnEmbeddingConfig {
        column: "i_item_desc".to_string(),
        model: "openai_embeddings".to_string(),
        primary_keys: Some(vec!["i_item_sk".to_string()]),
        chunking: None,
    }];

    let app = AppBuilder::new("text-to-sql")
        .with_dataset(get_taxi_trips_dataset())
        .with_dataset(ds_tpcds_item)
        .with_embedding(get_openai_embeddings(
            Some("text-embedding-3-small"),
            "openai_embeddings",
        ))
        .build();

    let rt = Arc::new(Runtime::builder().with_app(app).build().await);

    let (_tracing, trace_provider) = init_tracing_with_task_history(None, &rt);

    tokio::select! {
        () = tokio::time::sleep(std::time::Duration::from_secs(30)) => {
            return Err(anyhow::anyhow!("Timed out waiting for components to load"));
        }
        () = rt.load_components() => {}
    }

    runtime_ready_check(&rt).await;

    verify_sql_query_chat_completion(Arc::clone(&rt), &trace_provider).await?;
    verify_similarity_search_chat_completion(Arc::clone(&rt), &trace_provider).await?;

    Ok(())
}

/// Verifies that the model correctly uses the SQL tool to process user query and return the result
async fn verify_sql_query_chat_completion(
    rt: Arc<Runtime>,
    trace_provider: &TracerProvider,
) -> Result<(), anyhow::Error> {
    let model =
        get_openai_chat_model(Arc::clone(&rt), "gpt-4o-mini", "openai_model", "auto").await?;
    let req = CreateChatCompletionRequestArgs::default()
            .messages(vec![ChatCompletionRequestSystemMessageArgs::default()
                .content("You are an assistant that responds to queries by providing only the requested data values without extra explanation.".to_string())
                .build()?
                .into(),ChatCompletionRequestUserMessageArgs::default()
                .content("Provide the total number of records in the taxi trips dataset. If known, return a single numeric value.".to_string())
                .build()?
                .into()])
            .build()?;

    let task_start_time = std::time::SystemTime::now();
    let response = model.chat_request(req).await?;

    insta::assert_snapshot!(
        "chat_1_response_choices",
        format!("{:#?}", response.choices)
    );

    let _ = trace_provider.force_flush();

    let tasks = get_executed_tasks(&rt, task_start_time.into()).await?;

    assert!(
        tasks.iter().any(|t| { t.0 == "tool_use::list_datasets" }),
        "Expected 'tool_use::list_datasets' task to be executed"
    );
    assert!(
        tasks.iter().any(|t| { t.0 == "tool_use::sql" }),
        "Expected 'tool_use::sql' task to be executed"
    );
    assert!(
        tasks.iter().any(|t| { t.0 == "sql_query" }),
        "Expected 'sql_query' task to be executed"
    );

    let ai_completion_task =
        tasks
            .iter()
            .find(|t| t.0 == "ai_completion")
            .ok_or(anyhow::anyhow!(
                "Expected 'ai_completion' task to be executed"
            ))?;

    // ai_completion input message is deterministic - based on available tools, app configuration, and the input message
    insta::assert_snapshot!(
        "chat_1_ai_completion_input",
        pretty_json_str(&ai_completion_task.1)?
    );

    Ok(())
}

/// Verifies that the model correctly uses similirity search tool to process user query and return the result
async fn verify_similarity_search_chat_completion(
    rt: Arc<Runtime>,
    trace_provider: &TracerProvider,
) -> Result<(), anyhow::Error> {
    let model =
        get_openai_chat_model(Arc::clone(&rt), "gpt-4o-mini", "openai_model", "auto").await?;
    let req = CreateChatCompletionRequestArgs::default()
                .messages(vec![ChatCompletionRequestSystemMessageArgs::default()
                    .content("You are an assistant that responds to queries by providing only the requested data values without extra explanation.".to_string())
                    .build()?
                    .into(),ChatCompletionRequestUserMessageArgs::default()
                    .content("Find information about vehicles and journalists".to_string())
                    .build()?
                    .into()])
                .build()?;

    let task_start_time = std::time::SystemTime::now();
    let mut response = model.chat_request(req).await?;

    // Verify model correctly found records in the dataset
    assert!(response.choices[0]
        .message
        .content
        .clone()
        .unwrap_or_default()
        .contains("there just big vehicles. Journalists"));

    // response contains distance metrics that can vary so we replace content with a placeholder for snapshotting
    response.choices.iter_mut().for_each(|c| {
        c.message.content = Some("__placeholder__".to_string());
    });

    insta::assert_snapshot!(
        "chat_2_response_choices",
        format!("{:#?}", response.choices)
    );

    // ensure all spans are exported into task_history
    let _ = trace_provider.force_flush();

    let tasks = get_executed_tasks(&rt, task_start_time.into()).await?;

    let document_similarity_task = tasks
        .iter()
        .find(|t| t.0 == "tool_use::document_similarity")
        .ok_or(anyhow::anyhow!(
            "Expected 'document_similarity' task to be executed"
        ))?;

    // ai_completion input message is deterministic - based on available tools, app configuration, and the input message
    insta::assert_snapshot!(
        "chat_2_document_similarity_input",
        pretty_json_str(&document_similarity_task.1)?
    );

    Ok(())
}

fn get_openai_model(model: impl Into<String>, name: impl Into<String>) -> Model {
    let mut model = Model::new(format!("openai:{}", model.into()), name);
    model.params.insert(
        "openai_api_key".to_string(),
        "${ secrets:SPICE_OPENAI_API_KEY }".into(),
    );
    model
}

async fn get_openai_chat_model(
    rt: Arc<Runtime>,
    model: impl Into<String>,
    name: impl Into<String>,
    tools: impl Into<String>,
) -> Result<Box<dyn Chat>, anyhow::Error> {
    let mut model_with_tools = get_openai_model(model, name);
    model_with_tools
        .params
        .insert("tools".to_string(), tools.into().into());

    let model_secrets = get_params_with_secrets(&model_with_tools.params, &rt).await;
    try_to_chat_model(&model_with_tools, &model_secrets, rt)
        .await
        .map_err(anyhow::Error::from)
}

fn get_openai_embeddings(model: Option<impl Into<String>>, name: impl Into<String>) -> Embeddings {
    let mut embedding = match model {
        Some(model) => Embeddings::new(format!("openai:{}", model.into()), name),
        None => Embeddings::new("openai", name),
    };
    embedding.params.insert(
        "openai_api_key".to_string(),
        "${ secrets:SPICE_OPENAI_API_KEY }".into(),
    );
    embedding
}
