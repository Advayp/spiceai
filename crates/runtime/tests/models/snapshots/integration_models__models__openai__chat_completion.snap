---
source: crates/runtime/tests/models/openai.rs
expression: "normalize_chat_completion_response(response, false)"
snapshot_kind: text
---
{
  "choices": [
    {
      "finish_reason": "stop",
      "index": 0,
      "logprobs": null,
      "message": {
        "content": "10",
        "function_call": null,
        "refusal": null,
        "role": "assistant",
        "tool_calls": null
      }
    }
  ],
  "created": "created_val",
  "id": "id_val",
  "model": "openai_model",
  "object": "chat.completion",
  "service_tier": null,
  "system_fingerprint": "system_fingerprint_val",
  "usage": {
    "completion_tokens": "completion_tokens_val",
    "prompt_tokens": "prompt_tokens_val",
    "total_tokens": "total_tokens_val"
  }
}
